src/v/cluster/archival/ntp_archiver_service.cc

```diff
@@ -802,6 +826,206 @@ ss::future<> ntp_archiver::upload_until_term_change() {
     }
 }
 
+// Target segment size + min acceptable size
+struct segment_size_limits {
+    size_t target;
+    size_t lowest;
+};
+
+static segment_size_limits get_segment_size_limits() {
+    auto opt_target
+      = config::shard_local_cfg().cloud_storage_segment_size_target();
+    auto opt_min = config::shard_local_cfg().cloud_storage_segment_size_min();
+    if (opt_target == std::nullopt) {
+        // Use defaults
+        auto log_segment_size = config::shard_local_cfg().log_segment_size();
+        auto min_size = log_segment_size * 100 / 80;
+        return segment_size_limits{
+          .target = log_segment_size, .lowest = min_size};
+    }
+    return segment_size_limits{
+      .target = opt_target.value(),
+      .lowest = opt_min.value_or(opt_target.value() * 100 / 80)};
+}
+
+ss::future<> ntp_archiver::upload_until_term_change() {
+    vassert(_ops, "The method can't be called in legacy mode");
+
+    if (!_feature_table.local().is_active(
+          features::feature::cloud_storage_manifest_format_v2)) {
+        vlog(
+          _rtclog.warn,
+          "Cannot operate upload loop during upgrade, not all nodes "
+          "are upgraded yet.  Waiting...");
+        co_await _feature_table.local().await_feature(
+          features::feature::cloud_storage_manifest_format_v2, _as);
+        vlog(
+          _rtclog.info, "Upgrade complete, proceeding with the upload loop.");
+
+        // The cluster likely needed a bunch of restarts in order to
+        // reach this point, which means that leadership may have been
+        // transferred away (hence the explicit check).
+        if (!may_begin_uploads()) {
+            co_return;
+        }
+    }
+
+    // Before starting, upload the manifest if needed.  This makes our
+    // behavior more deterministic on first start (uploading the empty
+    // manifest) and after unclean leadership changes (flush dirty manifest
+    // as soon as we can, rather than potentially waiting for segment
+    // uploads).
+    {
+        auto units = co_await ss::get_units(_uploads_active, 1);
+        co_await maybe_upload_manifest(upload_loop_prologue_ctx_label);
+        co_await flush_manifest_clean_offset();
+    }
+
+    // This variable is set by the previous iteration of the upload loop.
+    // It's only used if the legacy mode is disabled.
+    std::optional<std::chrono::milliseconds> carried_over_backoff;
+
+    while (may_begin_uploads()) {
+        // Handle backoff before acquiring units from the semaphore. This
+        // will allow housekeeping to continue working.
+        if (carried_over_backoff.has_value()) {
+            co_await ss::sleep_abortable(carried_over_backoff.value(), _as);
+        }
+
+        // Hold sempahore units to enable other code to know that we are in
+        // the process of doing uploads + wait for us to drop out if they
+        // e.g. set _paused.
+        vassert(!_paused, "may_begin_uploads must ensure !_paused");
```

Comment:
It seems possible that a leadership transfer could happen as we're sleeping above, and _paused would be set to true. Maybe we should check `may_begin_uploads()` again after sleeping